actors/evm/src/interpreter/precompiles/bls12_381/g1_add.rs

```diff
@@ -0,0 +1,256 @@
+use crate::interpreter::{
+    System,
+    precompiles::{PrecompileContext, PrecompileError, PrecompileResult},
+};
+use fil_actors_runtime::runtime::Runtime;
+
+use crate::interpreter::precompiles::bls_util::{
+    G1_ADD_INPUT_LENGTH, PADDED_G1_LENGTH, encode_g1_point, extract_g1_input, is_infinity,
+    p1_from_affine, p1_to_affine,
+};
+
+use blst::{blst_p1, blst_p1_add_or_double_affine, blst_p1_affine};
+
+/// **BLS12_G1ADD Precompile**
+///
+/// Implements G1 point addition according to [EIP-2537](https://eips.ethereum.org/EIPS/eip-2537).
+pub fn bls12_g1add<RT: Runtime>(
+    _: &mut System<RT>,
+    input: &[u8],
+    _: PrecompileContext,
+) -> PrecompileResult {
+    if input.len() != G1_ADD_INPUT_LENGTH {
+        return Err(PrecompileError::IncorrectInputSize);
+    }
+
+    // Split the input bytes into two segments representing each G1 point.
+    let a_bytes = &input[..PADDED_G1_LENGTH];
+    let b_bytes = &input[PADDED_G1_LENGTH..];
+
+    // Convert the input bytes to their corresponding BLST affine representations.
+    let a_aff = extract_g1_input(a_bytes, false)?;
+    let b_aff = extract_g1_input(b_bytes, false)?;
+
+    // If either point is at infinity, return the other point as the result.
+    if is_infinity(&a_aff) {
+        return Ok(encode_g1_point(&b_aff));
+    }
+    if is_infinity(&b_aff) {
+        return Ok(encode_g1_point(&a_aff));
+    }
+
+    // Perform the addition in Jacobian coordinates, then convert back to affine.
+    let result_aff = p1_add_affine(&a_aff, &b_aff);
+    Ok(encode_g1_point(&result_aff))
+}
+
+/// Adds two G1 points given in affine form and returns the result in affine form.
+///
+/// This method is safe for point doubling since it allows `a` and `b` to be identical.
+#[inline]
+pub(super) fn p1_add_affine(a: &blst_p1_affine, b: &blst_p1_affine) -> blst_p1_affine {
+    let a_jacobian = p1_from_affine(a);
+    let sum_jacobian = p1_add_or_double(&a_jacobian, b);
+    p1_to_affine(&sum_jacobian)
+}
+
+/// Adds a G1 point in Jacobian coordinates and a G1 point in affine form.
+///
+/// # Safety
+///
+/// All inputs are assumed valid due to earlier checks.
```

Comment:
If this function relies on the caller making checks required for _safety_, it needs to be marked unsafe. But, from what I can tell, this comment is simply incorrect and this function _doesn't_ make any assumptions about safety (e.g., the parameters are pointers).

Am I missing something?

---

Anyways... please remove/rewrite this comment regardless.

> All inputs are assumed valid due to earlier checks.

Who's responsibility is that? What does valid mean? Where are these checks? wHaT Do YoU wAnT Me tO dOOOooooOOOoOoo!

Reply:
After spending some time understanding rust safety I share @Stebalien's confusion on this comment.  The input params are safe immutable rust pointers owned by this code.  And the structs are declared as `#[repr(C)]` in BLST [here](https://github.com/supranational/blst/blob/master/bindings/rust/src/bindings.rs#L479) and [here](https://github.com/supranational/blst/blob/master/bindings/rust/src/bindings.rs#L429).

Reply:
Continuing the discussion from the fp2_to_g2 thread we should either 
1) remove this comment asserting that any inputs to this function are ok for rust safety guarantees
2) identify the invariants that must be satisfied concerning the validity of p and p_affine in order for the unsafe block to not violate rust memory safety guarantees and then label this function unsafe.